interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java

```diff
@@ -0,0 +1,336 @@
+
+  private class Tester {
+    private final SimpleRequest simpleRequest = SimpleRequest.newBuilder()
+        .setResponseSize(314159)
+        .setResponseType(PayloadType.COMPRESSABLE)
+        .setPayload(Payload.newBuilder()
+            .setBody(ByteString.copyFrom(new byte[271828])))
+        .build();
+
+    public void rstStreamAfterHeader() throws Exception {
+      try {
+        blockingStub.unaryCall(simpleRequest);
+        fail("Expected call to fail");
+      } catch (StatusRuntimeException ex) {
+        assertEquals(Status.Code.UNAVAILABLE, ex.getStatus().getCode());
+      }
+    }
+
+    public void rstStreamAfterData() throws Exception {
+      try {
+        blockingStub.unaryCall(simpleRequest);
+        fail("Expected call to fail");
+      } catch (StatusRuntimeException ex) {
+        assertEquals(Status.Code.UNAVAILABLE, ex.getStatus().getCode());
+      }
+
+      // Use async stub to verify data was received.
+      StreamRecorder<SimpleResponse> responseObserver = StreamRecorder.create();
+      asyncStub.unaryCall(simpleRequest, responseObserver);
+      responseObserver.awaitCompletion();
+      assertEquals(1, responseObserver.getValues().size());
+      assertEquals(Status.Code.UNAVAILABLE,
+          Status.fromThrowable(responseObserver.getError()).getCode());
+    }
+
+    public void rstStreamDuringData() throws Exception {
+      try {
+        blockingStub.unaryCall(simpleRequest);
+        fail("Expected call to fail");
+      } catch (StatusRuntimeException ex) {
+        assertEquals(Status.Code.UNAVAILABLE, ex.getStatus().getCode());
+      }
+
+      // Use async stub to verify data was not received.
+      StreamRecorder<SimpleResponse> responseObserver = StreamRecorder.create();
+      asyncStub.unaryCall(simpleRequest, responseObserver);
+      responseObserver.awaitCompletion();
+      assertEquals(0, responseObserver.getValues().size());
+      assertEquals(Status.Code.UNAVAILABLE,
+          Status.fromThrowable(responseObserver.getError()).getCode());
+    }
+
+    public void goAway() throws Exception {
+      blockingStub.unaryCall(simpleRequest);
+      blockingStub.unaryCall(simpleRequest);
+    }
+
+    public void ping() throws Exception {
+      blockingStub.unaryCall(simpleRequest);
+    }
+
+    public void maxStreams() throws Exception {
+      final int numThreads = 10;
+
+      // Preliminary call to ensure MAX_STREAMS setting is received by the client.
+      blockingStub.unaryCall(simpleRequest);
+
+      ListeningExecutorService threadpool =
+          MoreExecutors.listeningDecorator(newFixedThreadPool(numThreads));
+      List<ListenableFuture<?>> workerFutures = new ArrayList<ListenableFuture<?>>();
+      ManagedChannel channel = createChannel();
+      for (int i = 0; i < numThreads; i++) {
+        workerFutures.add(threadpool.submit(new MaxStreamsWorker(channel, i, simpleRequest)));
+      }
+      ListenableFuture<?> f = Futures.allAsList(workerFutures);
+      f.get();
```

Comment:
Warning, if one of the futures fails, the rest of them will not be forced.  (resolved?)  Does this code cope with some futures not completing?